apps/prairielearn/src/question-servers/freeform.js

```diff
@@ -1766,29 +1819,38 @@ export async function parse(submission, variant, question, course) {
   });
 }
 
+/**
+ * @param {import('../lib/db-types.js').Submission} submission
+ * @param {import('../lib/db-types.js').Variant} variant
+ * @param {import('../lib/db-types.js').Question} question
+ * @param {import('../lib/db-types.js').Course} question_course
+ * @returns {import('./types.js').QuestionServerReturnValue<import('./types.js').GradeResultData>}
+ */
 export async function grade(submission, variant, question, question_course) {
   return instrumented('freeform.grade', async () => {
     debug('grade()');
     if (variant.broken_at) throw new Error('attempted to grade broken variant');
     if (submission.broken) throw new Error('attempted to grade broken submission');
 
     const context = await getContext(question, question_course);
+
+    /** @satisfies {import('./types.js').ExecutionData} */
     let data = {
       // Note that `params` and `true_answer` can change during `parse()`, so we
       // use the submission's values when grading.
-      params: submission.params,
-      correct_answers: submission.true_answer,
-      submitted_answers: submission.submitted_answer,
-      format_errors: submission.format_errors,
+      params: submission.params ?? {},
+      correct_answers: submission.true_answer ?? {},
+      submitted_answers: submission.submitted_answer ?? {},
+      format_errors: submission.format_errors ?? {},
```

Comment:
In practice, these should never be null, but they're required to satisfy the `ExecutionData` type here.

Reply:
I wonder if it's worth modeling `ExecutionData` with specializations from each phase. Probably not.

Reply:
I don't think we get much value from that. Declaring `data` with `satisfies ExecutionData` + the generics added to functions that process data mean we're already pretty much as safe as can be.

Reply:
Agreed.